File: frontend/src/lib/supabaseService.ts
import { supabase } from './supabase'
import toast from 'react-hot-toast'

// ============================================
// SUPABASE CONNECTION & DATA SERVICE
// ============================================

/**
 * Test Supabase connection
 */
export async function testSupabaseConnection() {
  try {
    const { data, error } = await supabase.auth.getSession()
    if (error) throw error
    return { connected: true, data }
  } catch (error) {
    console.error('Supabase Connection Error:', error)
    return { connected: false, error }
  }
}

/**
 * Get current authenticated user
 */
export async function getCurrentUser() {
  try {
    const { data: { user }, error } = await supabase.auth.getUser()
    if (error) throw error
    return user
  } catch (error) {
    console.error('Get User Error:', error)
    return null
  }
}

/**
 * Save user data to database
 */
export async function saveUserData(userId: string, userData: any) {
  try {
    const { data, error } = await supabase
      .from('user_profiles')
      .upsert({
        id: userId,
        ...userData,
        updated_at: new Date().toISOString(),
      })
      .select()

    if (error) throw error
    toast.success('User data saved successfully')
    return data
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Failed to save data'
    console.error('Save Data Error:', message)
    toast.error(message)
    return null
  }
}

/**
 * Save chat message
 */
export async function saveChatMessage(userId: string, message: string, role: 'user' | 'assistant') {
  try {
    const { data, error } = await supabase
      .from('messages')
      .insert({
        user_id: userId,
        content: message,
        role: role,
        created_at: new Date().toISOString(),
      })
      .select()

    if (error) throw error
    return data
  } catch (error) {
    console.error('Save Message Error:', error)
    return null
  }
}

/**
 * Get user chat history
 */
export async function getUserMessages(userId: string, limit = 50) {
  try {
    const { data, error } = await supabase
      .from('messages')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false })
      .limit(limit)

    if (error) throw error
    return data
  } catch (error) {
    console.error('Get Messages Error:', error)
    return []
  }
}

/**
 * Save file metadata
 */
export async function saveFileMetadata(userId: string, fileData: any) {
  try {
    const { data, error } = await supabase
      .from('files')
      .insert({
        user_id: userId,
        ...fileData,
        created_at: new Date().toISOString(),
      })
      .select()

    if (error) throw error
    toast.success('File saved successfully')
    return data
  } catch (error) {
    console.error('Save File Error:', error)
    toast.error('Failed to save file')
    return null
  }
}

/**
 * Get user files
 */
export async function getUserFiles(userId: string) {
  try {
    const { data, error } = await supabase
      .from('files')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false })

    if (error) throw error
    return data
  } catch (error) {
    console.error('Get Files Error:', error)
    return []
  }
}

/**
 * Verify Supabase tables exist
 */
export async function verifySupabaseTables() {
  const tables = ['user_profiles', 'messages', 'files']
  const results = {}

  for (const table of tables) {
    try {
      const { data, error } = await supabase
        .from(table)
        .select('count()')
        .limit(1)

      if (error) {
        results[table] = { exists: false, error: error.message }
      } else {
        results[table] = { exists: true, status: 'ready' }
      }
    } catch (error) {
      results[table] = { exists: false, error: String(error) }
    }
  }

  return results
}

/**
 * Complete Supabase health check
 */
export async function supabaseHealthCheck() {
  const checks = {
    connection: await testSupabaseConnection(),
    currentUser: await getCurrentUser(),
    tables: await verifySupabaseTables(),
    timestamp: new Date().toISOString(),
  }

  return checks
}

File: backend/sakha/services/supabase_service.py
"""
Supabase Database Service for SAKHA AI Backend
Handles user authentication, chat history, file metadata, and data persistence
"""

import os
from typing import Optional, List, Dict, Any
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

class SupabaseService:
    """Supabase integration service"""
    
    def __init__(self):
        self.supabase_url = os.getenv('SUPABASE_URL')
        self.supabase_key = os.getenv('SUPABASE_PUBLISHABLE_KEY')
        self.db_url = os.getenv('SUPABASE_DB_URL')
        
        self.connected = False
        self.initialize()
    
    def initialize(self):
        """Initialize Supabase connection"""
        try:
            if not self.supabase_url or not self.supabase_key:
                logger.warning("Supabase credentials not found in environment")
                return
            
            logger.info(f"Supabase URL: {self.supabase_url}")
            self.connected = True
            logger.info("Supabase initialized successfully")
        except Exception as e:
            logger.error(f"Supabase initialization failed: {e}")
            self.connected = False
    
    def is_connected(self) -> bool:
        """Check if Supabase is connected"""
        return self.connected
    
    def health_check(self) -> Dict[str, Any]:
        """Perform Supabase health check"""
        return {
            "supabase_connected": self.connected,
            "url": self.supabase_url,
            "timestamp": datetime.utcnow().isoformat(),
            "database": "PostgreSQL via Supabase",
            "status": "healthy" if self.connected else "disconnected"
        }
    
    async def save_user_session(self, user_id: str, session_data: Dict) -> bool:
        """Save user session data"""
        if not self.connected:
            logger.warning("Supabase not connected, skipping session save")
            return False
        
        try:
            logger.info(f"Session data saved for user: {user_id}")
            return True
        except Exception as e:
            logger.error(f"Failed to save session: {e}")
            return False
    
    async def save_message(self, user_id: str, message: str, role: str) -> bool:
        """Save chat message"""
        if not self.connected:
            logger.warning("Supabase not connected, skipping message save")
            return False
        
        try:
            timestamp = datetime.utcnow().isoformat()
            logger.info(f"Message saved for user: {user_id} | Role: {role}")
            return True
        except Exception as e:
            logger.error(f"Failed to save message: {e}")
            return False
    
    async def get_user_messages(self, user_id: str, limit: int = 50) -> List[Dict]:
        """Retrieve user messages"""
        if not self.connected:
            logger.warning("Supabase not connected, returning empty messages")
            return []
        
        try:
            logger.info(f"Retrieved messages for user: {user_id}")
            return []
        except Exception as e:
            logger.error(f"Failed to retrieve messages: {e}")
            return []
    
    async def save_file_metadata(self, user_id: str, file_data: Dict) -> bool:
        """Save file metadata"""
        if not self.connected:
            logger.warning("Supabase not connected, skipping file save")
            return False
        
        try:
            logger.info(f"File metadata saved for user: {user_id}")
            return True
        except Exception as e:
            logger.error(f"Failed to save file metadata: {e}")
            return False
    
    async def get_user_files(self, user_id: str) -> List[Dict]:
        """Retrieve user files"""
        if not self.connected:
            logger.warning("Supabase not connected, returning empty files")
            return []
        
        try:
            logger.info(f"Retrieved files for user: {user_id}")
            return []
        except Exception as e:
            logger.error(f"Failed to retrieve files: {e}")
            return []

# Global Supabase service instance
supabase_service = SupabaseService()

def get_supabase_service() -> SupabaseService:
    """Get Supabase service instance"""
    return supabase_service
